docs/src/faq.md

```diff
@@ -0,0 +1,402 @@
+```@meta
+CurrentModule = Enzyme
+DocTestSetup = quote
+    using Enzyme
+end
+```
+
+# Frequently asked questions
+
+## Implementing pullbacks
+
+In combined reverse mode, Enzyme's [`autodiff`](@ref) function can only handle functions with scalar output (this is not true for split reverse mode, aka `autodiff_thunk`).
+To implement pullbacks (back-propagation of gradients/tangents) for array-valued functions, use a mutating function that returns `nothing` and stores it's result in one of the arguments, which must be passed wrapped in a [`Duplicated`](@ref).
+
+Given a function `mymul!` that performs the equivalent of `R = A * B` for matrices `A` and `B`, and given a gradient (tangent) `∂z_∂R`, we can compute `∂z_∂A` and `∂z_∂B` like this:
+
+```@example pullback
+using Enzyme, Random
+
+function mymul!(R, A, B)
+    @assert axes(A,2) == axes(B,1)
+    @inbounds @simd for i in eachindex(R)
+        R[i] = 0
+    end
+    @inbounds for j in axes(B, 2), i in axes(A, 1)
+        @inbounds @simd for k in axes(A,2)
+            R[i,j] += A[i,k] * B[k,j]
+        end
+    end
+    nothing
+end
+
+Random.seed!(1234)
+A = rand(5, 3)
+B = rand(3, 7)
+
+R = zeros(size(A,1), size(B,2))
+∂z_∂R = rand(size(R)...)  # Some gradient/tangent passed to us
+∂z_∂R0 = copyto!(similar(∂z_∂R), ∂z_∂R)  # exact copy for comparison
+
+∂z_∂A = zero(A)
+∂z_∂B = zero(B)
+
+Enzyme.autodiff(Reverse, mymul!, Const, Duplicated(R, ∂z_∂R), Duplicated(A, ∂z_∂A), Duplicated(B, ∂z_∂B))
+```
+
+Now we have:
+
+```@example pullback
+R ≈ A * B            &&
+∂z_∂A ≈ ∂z_∂R0 * B'   &&  # equivalent to Zygote.pullback(*, A, B)[2](∂z_∂R)[1]
+∂z_∂B ≈ A' * ∂z_∂R0       # equivalent to Zygote.pullback(*, A, B)[2](∂z_∂R)[2]
+```
+
+Note that the result of the backpropagation is *added to* `∂z_∂A` and `∂z_∂B`, they act as accumulators for gradient information.
+
+## Identical types in `Duplicated`
+
+Enzyme forces `x` and `∂f_∂x` to have the same types when constructing objects of type `Duplicated`, `DuplicatedNoNeed`, `BatchDuplicated`, etc.
+This is not a mathematical requirement, but rather a guardrail to prevent user error.
+The memory locations of `x` and `∂f_∂x` are accessed in the same way by the differentiation code, so they should have the same data layout.
+Equality of types is an approximation of this condition.
+
+In some ways, type equality is too strict: two different types can have the same data layout.
```

Comment:
perhaps add, for example Tuple{Float64, Float64} or `struct Pair
  x::Float64
  y::Float64
` have the same datalayout Enzyme would produce the expected results when given a primal of a Pair and a shadow of the tuple of floats.

Reply:
My other example idea was a vector vs a matrix column view

Reply:
That's the one which really is a pain in the butt for me atm

Reply:
My code is riddled with stuff like
```julia
function DI.value_and_pushforward(backend::AutoForwardEnzyme, f, x, dx, extras::Nothing)
    dx_sametype = convert(typeof(x), dx)
    y, dy = autodiff(backend.mode, f, Duplicated, Duplicated(x, dx_sametype))
    return y, dy
end
```

Reply:
added the vector example

Reply:
Yeah so the fix for your matrix view stuff is to actually leverage the "advanced memory layout" stuff discussed below. Aka if an array just stores values 0...N, and a Matrix stores row 0 elems in 0...N, etc. You can technically force Enzyme to take a Duplicated(Array, Matrix) and it will store the derivative in row 0 (assuming the layout of julia arrays is what I expect). You can then force it to store in row 1, by passing in Duplicated(Array, unsafe_array(unsafe_pointer(Matrix) + N * sizeof(eltype))). [since then it will store the derivative data at offset 0 to the shadow, which is equivalent to starting at element N of your array, aka at Row 1, col 0].